File: iosSDK/DropthoughtSDK-Objc/Survey.m
//
//  Survey.m
//  DropthoughtSDK-Objc
//
//  Created by BCT-Barney on 2020/6/4.
//  Copyright © 2020 bct.tpe. All rights reserved.
//

#import <UIKit/UIKit.h>
#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>

#import "Survey.h"
#import "SurveyApplication.h"
#import "SurveyEmitter.h"
#import "SurveyViewController.h"


@interface Survey()
@property (nonatomic, strong) UIViewController *from;
@property (nonatomic, strong) SurveyApplication *app;
@property (nonatomic, strong) NSString *apiKey;
@end

@implementation Survey

RCT_EXPORT_MODULE();

+ (instancetype)sharedInstance {
    static Survey *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[Survey alloc] init];
    });
    return instance;
}

- (void)initSurvey:(NSDictionary *)launchOptions apiKey:(NSString *)apiKey {
    self.app = [[SurveyApplication alloc] init];
    [self.app setupBridge:launchOptions];
    self.apiKey = apiKey;
}

- (void)setupAPIKey:(NSString *)apiKey {
    self.apiKey = apiKey;
}

- (void)present:(UIViewController *)from surveyId:(NSString *)surveyId {
    self.from = from;

    NSDictionary *initialProperties = @{ @"apiKey" : self.apiKey, @"surveyId": surveyId };
    RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:self.app.bridge moduleName:@"dropthought-sdk" initialProperties:initialProperties];
    rootView.frame = [UIScreen mainScreen].bounds;

    SurveyViewController *vc = [[SurveyViewController alloc] init];
    vc.view = rootView;
    vc.modalPresentationStyle = UIModalPresentationFullScreen;

    [self.from presentViewController:vc animated:YES completion:NULL];
}

- (void)sendUploadOfflineFeedbacksEvent {
    [self.app.bridge enqueueJSCall:@"RCTDeviceEventEmitter" method:@"emit" args:@[@"UploadQueuedFeedback", @{@"apiKey": self.apiKey}] completion:NULL];
}


RCT_EXPORT_METHOD(dismiss)
{
    dispatch_async(dispatch_get_main_queue(), ^{
        [[[Survey sharedInstance] from] dismissViewControllerAnimated:YES completion:nil];
    });
}
@end

File: iosSDK/DropthoughtSDK/Survey.swift
//
//  Survey.swift
//  RNSDK DemoApp
//
//  Created by BCT-Barney on 2020/5/7.
//  Copyright © 2020 bct.tpe. All rights reserved.
//

import UIKit
import React

public class Survey: NSObject {
    public static func present(_ from: UIViewController, apiKey: String, surveyId: String) {
        printBundles()

        guard let bundleURL = Bundle(for: Survey.self).url(forResource: "main", withExtension: "bundle"),
            let jsBundleLocation = Bundle(url: bundleURL)?.url(forResource: "main", withExtension: "jsbundle") else {
            return
        }
        let data = ["apiKey": apiKey, "surveyId": surveyId]
        let rootView = RCTRootView(bundleURL: jsBundleLocation, moduleName: "dropthought-sdk", initialProperties: data, launchOptions: nil)
        
        let viewController = UIViewController()
        rootView.frame = UIScreen.main.bounds
        viewController.view.addSubview(rootView)
        viewController.modalPresentationStyle = .fullScreen
        from.present(viewController, animated: true, completion: nil)
    }

    static func printBundles() {
        let docsPath = Bundle(for: Survey.self).resourcePath!
        let fileManager = FileManager.default

        do {
            let docsArray = try fileManager.contentsOfDirectory(atPath: docsPath)
            print(docsArray)
        } catch {
            print(error)
        }
    }
}
